src/ray/raylet/scheduling_queue.cc

```diff
@@ -195,27 +195,53 @@ void SchedulingQueue::FilterState(std::unordered_set<TaskID> &task_ids,
   }
 }
 
-std::vector<Task> SchedulingQueue::RemoveTasks(std::unordered_set<TaskID> &task_ids) {
+std::pair<std::vector<Task>, std::vector<TaskState>> SchedulingQueue::RemoveTasks(
+    std::unordered_set<TaskID> &task_ids) {
```

Comment:
How about passing `std::vector<TaskState> *removed_states` into the method and if it is not nullptr, then we will return the task states there?